### ISSUE TITLE
feat(migrate): have desktop coordinate backend migrations needed to get desktop app to work

### ISSUE THREAD

**Issue by ramfox** [qri-io/desktop#566]
Let's use the qri config, and electron's `ipcMain` and `ipcRender` events to coordinate backend migrations.

We can also use `qri version` to ensure the user is running the correct backend (if "-dev" is present, or `qri connect` is already running, we should ignore this)

### why fail based on `qri version` and `config`?
Qri Desktop, in production, has an implicit contract with the Qri Backend that the WebSocket and HTTP api's work in an expected way. We should ensure that, unless the user is using a `dev` version of qri, these expectations are preserved. 

### how to proceed

in `backend.js` file let's add `string lowestCompatibleBackend` and `number lowestCompatibleConfigRevision`.

in `maybeStartup` func, ping the `/health` endpoint. If we get a response, just return! (This ensures that is qri is  already running, we don't interfere)

in `launchProcess` func:

If the qri version on the PATH is lower than the `lowestCompatibleBackend`, we should send an "incompatible-backend" event (with the incompatible version string) to the main process. (Unless the given qri version has a `-dev` version ending, then just proceed with attempting to connect as normal.)

If the given config revision is lower than the expected, send a "migrating backend" event to the main process, and start qri using `qri connect --migration`. If that fails, send a "migration failed" event to the main process.

The main process should handle these "incompatible-backend", "migrating-backend", "migration-failed" events, and use the `window.webContents.send` func to pass these events to the render processes.

The render process should handle these events.

### UI reducer and the render process event handlers

new type in `/reducer/ui.ts`:
```
export type bootupComponent:  'loading' | 'migrating' | 'migrationFailure' | string
```

UI should have a new field `bootupComponent: bootupComponentType`, which by default is 'loading'

### new components

Create three new components:
- MigratingBackend -> informs the user that we are running migrations on qri and this may take a minute
- MigrationFailed -> informs the user that the migration failed and gives instructions on how to send a crash report
- IncompatibleBacken -> displays expected version and version that caused the error

(detailed in separate issue: https://github.com/qri-io/desktop/issues/567)

### App component

The ipcRender process should have a handler that listens for "incompatible-backend", "migrating-backend", and "migration-failed" events. In each case, send an action to update the UI reducer's bootupComponent field. (in the case of "incompatible-backend", set the field to the incompatible version number). These events should be registered when we use the App component.

In `App.tsx`, add a new prop `bootupComponent`, and pass it the UI.bootupComponent. If we are `loading`, switch on `bootupComponentType` (if not 'loading', 'migrating' or 'migrationFailure', default should be `bootupComponent` is the incompatible version number to pass to a `IncompatibleBackend` component).

**Comment by uhLeeshUh** [comment]
Sounds good! Quick question @ramfox -- I'm a tad confused by why you call out that if the `qri version` has `-dev` then we'll ignore the migration step. Would a `qri version` only have that dev flag if it's the latest version? So is this a shortcut to know that the version the user has must be compatible if it's inherently the latest?

**Comment by ramfox** [comment]
@uhLeeshUh I made that decision based on the fact that as devs we might be trying out odd setups that can have strange configs and behaviors. I'm going off the assumption that when someone is using a `-dev` version, they have agreed to the social contract that things may not work perfectly (and if they can fix it if it does).

Same thing goes for if someone is running `qri connect`. That is an advanced use-case for Qri Desktop, so we are going to assume you have some knowledge of what's going on, since you ran `qri connect` on purpose. Also, if someone is already running `qri connect` we can't auto-migrate anyways.

Your average user, however, did not agree to that social contract, and so we need to make sure that everything is aligned when those folks go to use the Qri Desktop app. Most of these checks are for them.

**Comment by ramfox** [comment]
So I didn't fully understand our `BackendProcess` class and how it was launching the backend. I thought, erroneously, that it was somehow an electron render process (which i now know is actually tied to the BrowserWindow). So, in order to get the granularity we want, let's catch any errors thrown and use the `window.webContents.send()` function to alert any listening render functions:

```js
      backendProcess.checkNoActiveBackendProcess()
        .then(backendProcess.checkNeedsMigrations()
        .then(backendProcess.launchProcess, (err) => {
          if (err.message === "backend-needs-migration") {
            window.webContents.send("migrating-backend")
            backendProcess.migrateAndLaunchProcess()
            return
          }
          throw err
        })
        .catch(err => {
          switch (err.message) {
            case "backend-already-running":
              window.webContents.send("backend-already-running")
            case "backend-dev":
              window.webContents.send("backend-dev")
            case "incompatible-backend":
              window.webContents.send("incompatible-backend", backendProcess.backendVer)
            case "error-launching-backend":
              window.webContents.send("error-launching-backend")
            default:
              log.error(err.message)
          }
        })
```

**Comment by ramfox** [comment]
Okay another redesign - we are now not ignoring if something has the `-dev` flag. Now the app will migrate/warn about incompatibility even if you have a version with the `-dev` suffix

**Comment by ramfox** [comment]
Final edit:
In dev mode the app doesn't build fast enough to get any of the events!

In order to ensure that the renderer is built before we start the process, let's have the `App` component emit an `app-fully-loaded` event once all the handlers that manage the coordination on the renderer's side get built!

In the main process, let's add an `app-fully-loaded` handler, and only launch the backend processes when this event has been emitted. Now we know for sure that the renderer will be there to handle the emitted migration/backend launching events.

Because reloading can trigger the `app-fully-loaded` event, we need to make sure that if we get a `backend-already-running` error, we `.close()` the created `BackendProcess`. In fact, any error that emits from the launching process should explicitly close the `BackendProcess`
